### PR TITLE
Improve UX for child record creation and management

### PR DESCRIPTION
**Pull request by stevecassidy**
# Feat: Improve UX for child record creation and management

## JIRA Ticket

[BSS-364](https://jira.csiro.com/browse/BSS-364)

## Description

BSS prefers a style for child records (buildings) where they are shown inline in the main form.  Current Fieldmark creates a link to the child record and shows the list of links instead.  Editing a child record jumps to a separate page for that form and saving comes back.

Fully embedding the child record form is probably too involved for v1 implementation but the experience of child records could be improved.

## Proposed Changes

- [x] Improve the layout of the RelatedRecordField to use the name of the child form (Building) in the action buttons
- [x] Derive the label on the add record button from the form name - instead of "Add child record" it could be "Add <FormName>" ("Add Building"). 
- [x] When child records exist, show a more detailed summary list, perhaps more like the updated record list in the notebook view with customisable fields shown
- [x] Update the 'publish' buttons on a child record to refer to the parent type 'Publish and return to Site' and 'Publish and New Building'
- [x] Use `RecordMetadata` in the child record list rather than `RecordLinkProps`, modify `listRecordMetadata` to return field values as well as metadata so we can include field values in the record listing

## How to Test

Update to the most recent Rapid Site Survey from github since it includes the summary fields configuration. 

Create a site record. Note the button to make new Buildings or add an existing building.   Create a new building, note the names on the buttons on that form.   Publish the building and return to the site, notice the fields shown in the building listing (currently 'building name' and 'building damage' but can be customised).  

Check that buildings are listed in the Site as expected. Add multiple buildings.

## Notes

This has exposed for me how awful the handling of records is in the app.  Too many similar types for records and usually the wrong one where you need to get some information like the type or field values.  Desperately need to refactor the interface to the records database to make it more orthogonal and less annoying.

## Checklist

- [x] I have confirmed all commits have been signed.
- [x] I have added JSDoc style comments to any new functions or classes.
- [x] Relevant documentation such as READMEs, guides, and class comments are updated.

### CODE REVIEW

**Comment by PeterBaker0** [comment]
I cant currently see the summary details in the buildings link section with newest notebook - not sure if I've set something up wrong

Notebook: https://github.com/nbic-bss/bss-survey-templates/blob/main/rapid-impact-survey/rapid-impact-survey.json

Screenshot
image

File: app/src/gui/components/record/formButton.tsx [hunk 1]
```diff
@@ -76,8 +80,34 @@ export default function FormButtonGroup(props: any) {
...
+
+        setLabels({
+          ...labels,
+          close: `Save and return to ${parentName}`,
+          new: `Publish and new ${record_type}`,
+        });
+      }
+    };
+    get();
+  }, []);
```
Comment (PeterBaker0) [inline app/src/gui/components/record/formButton.tsx:109]: might want to put project_id and relationship into the dependency array - seems weird that the relationship can't just pass through directly the record types rather than doing this async fetch thing - but is what it is

File: app/src/gui/components/record/relationships/RelatedInformation.tsx [hunk 2]
```diff
@@ -321,24 +337,29 @@ export async function get_RelatedFields_for_field(
   multiple: boolean,
   related_type_label: string | undefined,
   form_type: string | undefined,
-  realtion_type: string
+  relation_type: string
```
Comment (PeterBaker0) [inline app/src/gui/components/record/relationships/RelatedInformation.tsx:340]: It'd be worth refactoring this function to use a named object input rather than ordered params given how many there are. 

E.g. 

```
function getRelatedRecords({argA, argB} : {argA: typeA, argB: typeB}) {
...
}
```

File: app/src/gui/components/record/relationships/RelatedInformation.tsx [hunk 3]
```diff
@@ -675,15 +698,9 @@ function get_field_label(ui_specification: ProjectUIModel, field: string) {
...
   }
   try {
-    if (
-      ui_specification['fields'][field]['component-parameters'][
-        'InputLabelProps'
-      ]['label']
-    )
+    if (ui_specification['fields'][field]['component-parameters'].label)
       field_name =
-        ui_specification['fields'][field]['component-parameters'][
```
Comment (PeterBaker0) [inline app/src/gui/components/record/relationships/RelatedInformation.tsx:684]: What's the rationale for this change? I notice you've removed InputLabelProps?

File: app/src/gui/components/record/relationships/create_links/create_record_link.tsx [hunk 4]
```diff
@@ -90,14 +91,20 @@ export function CreateRecordLink(props: CreateRecordLinkProps) {
...
+    relatedRecords,
     handleChange,
-    relationshipLabel,
     SetSelectedRecord,
     selectedRecord,
     disabled,
     project_id,
   } = props;
+
+  console.log('CreateRecordLink', relatedRecords);
```
Comment (PeterBaker0) [inline app/src/gui/components/record/relationships/create_links/create_record_link.tsx:102]: want to remove for prod?

File: app/src/gui/components/record/relationships/field_level_links/datagrid.tsx [hunk 5]
```diff
@@ -36,15 +37,19 @@ import {
...
-import Checkbox from '@mui/material/Checkbox';
-import CheckCircleIcon from '@mui/icons-material/CheckCircle';
 import {gridParamsDataType} from '../record_links';
+import {RecordLinkProps} from '../types';
+import {getRecordInformation} from '../RelatedInformation';
+import {
+  getFieldLabel,
+  getSummaryFields,
+  getUiSpecForProject,
+} from '../../../../../uiSpecification';
```
Comment (PeterBaker0) [inline app/src/gui/components/record/relationships/field_level_links/datagrid.tsx:52]: this import path is pretty nuts haha

File: app/src/gui/components/record/relationships/field_level_links/datagrid.tsx [hunk 6]
```diff
@@ -190,122 +209,105 @@ export default function DataGridFieldLinksComponent(
...
+      const fn = async () => {
+        // get the record so we can display some fields
+        const rd = await getRecordInformation(props.child_record);
+        setRecordData(rd);
+
+        if (uiSpec && rd?.latest_record)
+          setDisplayFields(getSummaryFields(uiSpec, rd.latest_record.type));
+      };
+      fn();
+    }, [props.child_record]);
```
Comment (PeterBaker0) [inline app/src/gui/components/record/relationships/field_level_links/datagrid.tsx:240]: Might be worth starting to use React Query instead of useEffects to manage these async fetches. 

This merges the useState, the useEffect etc into one operation which includes loading, error by default so a bit neater/nicer.

File: app/src/gui/components/record/relationships/field_level_links/datagrid.tsx [hunk 7]
```diff
@@ -150,6 +158,17 @@ export default function DataGridFieldLinksComponent(
...
+  const [uiSpec, setUISpec] = React.useState<ProjectUIModel | null>(null);
+
+  useEffect(() => {
+    const get = async () => {
+      const u = await getUiSpecForProject(props.project_id);
+      setUISpec(u);
+    };
+
+    get();
+  }, [props.project_id, props.field_name]);
```
Comment (PeterBaker0) [inline app/src/gui/components/record/relationships/field_level_links/datagrid.tsx:170]: I'm a bit worried about performance/complexity. Shouldn't the ui spec be available already at this point given we are deep within a form? Why does it need to fetch it again?

File: app/src/gui/components/record/relationships/field_level_links/datagrid.tsx [hunk 8]
```diff
@@ -190,122 +209,105 @@ export default function DataGridFieldLinksComponent(
...
-              props.handleMakePreferred(
-                params.row.record_id,
-                event.target.checked
+    const [displayFields, setDisplayFields] = useState<Array<string>>([]);
+
+    useEffect(() => {
+      const fn = async () => {
+        // get the record so we can display some fields
+        const rd = await getRecordInformation(props.child_record);
+        setRecordData(rd);
```
Comment (PeterBaker0) [inline app/src/gui/components/record/relationships/field_level_links/datagrid.tsx:234]: Is there any error handling here? What happens if getRecordInformation fails e.g. due to permissions issue?

**Review by PeterBaker0 — COMMENTED** [review]
Overall I can see the frustration regarding record handling. My comments are suggestions on how some aspects could be improved. 

- don't like the over reliance on useEffect with startup async operations deep in the form - at this point the app should not need to be consulting the database for that information since the props should have that info at some point in the component tree. If you must do fetches etc would prefer to wrap with useQuery so we get error handling, load states, useEffect management for free.


Also wondering if the designer has been updated with these changes, or needs to be? And how we manage documentation for these kind of changes going forward? For now, tickets to capture those missing pieces should be fine.

I couldn't actually get the child relationship panel to load as you describe in my instance - we can discuss this more.

File: app/src/gui/components/record/relationships/RelatedInformation.tsx (hunk 3 above)
Comment (stevecassidy) [reply app/src/gui/components/record/relationships/RelatedInformation.tsx:684]: `InputLabelProps` was an old way to set a label on this and a few other fields, I've now standardised to `['component-parameters'].label` for all field types.

File: app/src/gui/components/record/formButton.tsx (hunk 1 above)
Comment (stevecassidy) [reply app/src/gui/components/record/formButton.tsx:109]: done

File: app/src/gui/components/record/relationships/RelatedInformation.tsx (hunk 2 above)
Comment (stevecassidy) [reply app/src/gui/components/record/relationships/RelatedInformation.tsx]: Reduced the number of required params here.

File: app/src/gui/components/record/relationships/create_links/create_record_link.tsx (hunk 4 above)
Comment (stevecassidy) [reply app/src/gui/components/record/relationships/create_links/create_record_link.tsx:102]: done

File: app/src/gui/components/record/relationships/field_level_links/datagrid.tsx (hunk 7 above)
Comment (stevecassidy) [reply app/src/gui/components/record/relationships/field_level_links/datagrid.tsx:174]: I agree, we should have a context provider or something.

File: app/src/gui/components/record/relationships/field_level_links/datagrid.tsx (hunk 6 above)
Comment (stevecassidy) [reply app/src/gui/components/record/relationships/field_level_links/datagrid.tsx]: Removed the need for a query here by getting the full record data before we get here.

File: app/src/gui/components/record/relationships/field_level_links/datagrid.tsx (hunk 8 above)
Comment (stevecassidy) [reply app/src/gui/components/record/relationships/field_level_links/datagrid.tsx]: Not doing that here any more.

**Comment by PeterBaker0** [comment]
image
I notice if I remove a link, then add it back via link existing record, that the list won't be updated until I leave that page and return

File: app/src/gui/components/record/relationships/RelatedInformation.tsx [hunk 9]
```diff
@@ -871,6 +891,14 @@ export async function addRecordLink(
...
       await upsertFAIMSData(child_record.project_id, new_doc);
     }
+    // here we are trusting that Record has enough of the fields of RecordMetadata
+    // for the purposes of the caller until such time as we rationalise the Record types
+    child_record_meta = latest_record as unknown as RecordMetadata;
+    // it's missing the HRID so grab it here
+    child_record_meta.hrid = getHRIDValue(
+      child_record_meta.record_id,
+      child_record_meta.data
+    );
```
Comment (PeterBaker0) [inline app/src/gui/components/record/relationships/RelatedInformation.tsx:901]: 🤦

**Review by PeterBaker0 — APPROVED** [review]
Approved pragmatically but man do records need a refactor! Thanks for trudging through this.

File: api/src/couchdb/initialise.ts [hunk 10]
```diff
@@ -235,8 +235,8 @@ export const initialiseAuthDb = async (db: PouchDB.Database): Promise<void> => {
   let initialised = false;
```
[the remark this answers is not in the document]
Comment (PeterBaker0) [reply api/src/couchdb/initialise.ts]: I fixed this

**Comment by PeterBaker0** [comment]
image

Not seeing the record get linked now - adding building reveals no buildings in the linked list

**Comment by PeterBaker0** [comment]
> image
> 
> Not seeing the record get linked now - adding building reveals no buildings in the linked list

Resolved - local build was broken

**Review by PeterBaker0 — APPROVED** [review]
See previous - issue resolved
